modules/store/src/reducer_creator.ts

```diff
@@ -40,15 +40,12 @@ export interface OnReducer<State, Creators extends readonly ActionCreator[]> {
  * ```
  */
 export function on<State, Creators extends readonly ActionCreator[]>(
-  ...args: [
-    ...creators: Creators,
-    reducer: OnReducer<State extends infer S ? S : never, Creators>
-  ]
+  ...args: [...creators: Creators, reducer: OnReducer<State, Creators>]
```

Comment:
While this fixes the issue of inferring `State` when `on` is used outside `createReducer`, I would expect that it breaks the deferred contextual inference which was intended by using the conditional type. Do the current tests ensure that `State` is inferred contextually from `on`'s usage in the parameters of a function like `createReducer`? Otherwise, I'd expect this could break some intended functionality.

Reply:
Specifically, this allowed for generic reducers: https://github.com/ngrx/platform/pull/2996/files#diff-b5b61b5c88ce4368769a58df0918602484c6c507e6a6dfad51e4746e470caf26R45

Reply:
In that last PR, it seems the effect of using the conditional type was not understood. Here's why it worked- https://stackoverflow.com/a/70129630/11087018.

I'm not immediately sure what the correct thing to do would be to support both use cases. Probably a good question for jcalz, the internet's StackOverflow Typescript wizard.

Reply:
Hey @david-shortman 
Yes, we have a test covering a generic reducer as well. 
There are number of test still broken, so this is not the final fix.

Reply:
Well, this was fun :) 

@david-shortman this should be now fixed with preserving generic reducers and all the other interesting cases that we were running into.

[Here's the playground link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgIILMA9iAShAEwFckoAeAFQD5kBvAKGWQAoBnMOSALmQuQB9kREAQgxQhAJQ8KAbnoBfehAAeAByxQwyUJFiIU+YqQoBPNRFZlGyAModIAGhsBhKBE6bWyVZBHf3OAIcABtTZHYoUABzAG0AXWcaBiZ3Y2geAHk8QhJoMntOCEdkNw8wLyp5JWV1TW1daHgkZGyjPPJCp1KfFT8CAI9gkDCIsCiQOMTkAEkQGGg0rpQAXjsHYt5kNeEAaxAsAHcQXv7vZeQAfln5xcILnmXkmzYNnj2D49OIf3Wiq5uCygSw2yEeGxKCECFSgrB4LmkvGq9HoAHpUbwAKK2Ci2egIHDsZA4XBZHLpcgfI4gEqxADkdOmtBgWCwPEiMQUNDWbBKcEk2xorHkBJARJJACYye1SGRmaz2eNObSGdMqcduSxWHyBSshSLCdoSQBmaW5WXytljCbRBQqxma3nIfmCiIGsVGvAAFm2Wp4lsVNrtzp4sQ5k3iuv1+MNxJAADEQAB1YBgAAWuF9OGYdJZWDpzu84eiJRJXsk8noogQITg7mQMGEGGwJxwBQhPV8PwGyECw1GxYSJTmQJB-zWT2YNgAdLO69E4chYrPp1Dyl54SU0h0zRT2-8u79QEC7ADbGDkCAIAA3aAlFzD27A+4bKjxeiImXQMwWKzqk6Hj2FzXCOdwEA8fzdC4VQotWtb1o2IDNjgyBrkUX6dFQU5MKAqbAHAITLI8zhMCuhI8BhP6WAUtLFvEVAJB+PDoJgJLmvktgwWiGIALR8fxAmCfxKLoliOIzAAcgA4lwXD0IhyEnGhkBST80DAAgGGUBcgHeFgABGABWEAYFhuGYARRG8MsAopEw9lMKKRLbqQvrKRAGHYQ53nmfhhEQjY3n2dmdJwHSJSvEUUZ0IFQX2e4YBEFAJzsP88hxfZCiSLFDkViiGUJUlJwudA6XeTUjT6C0ywAAp1mAACMdB5pcPAgEQAC2+mlUolXNCgtX1RKdD6XWbWdd1UCyEoTnaHmABCdYYW50IeexnQbHVWhNYIg1aBKWE5d5tB2kdDkhXm4Vahs0XMLQyArql3R5jwTVZZIJEZXGOajVAV2RZAt33Y9Ha-a9yDvWdyB5fQfUGJBKD3SAcAdRAgYxBD7rigmJxrBdrIFnARZKpMEVPWjCNA5eKMU3SrBYKjyDXgRRAQAW72VrNcaJsNeMgDml2FtaMRkzdrp3dTqM8HTDMoMzISs+zkgwzGHq9utK3lGte6TlDTC0MjUvIHSIDs59GU4Im5veTDXMlVAvOoatnl63Qhu06bEPW0Fls0lDKuiawaZYEQIQEMgmK4LgmS4KrznrcamvoetXkZbQYPG57p1fdzfsZQHGJByHYcR1HMdx9o9s+ms7laenY3C5MXKp3F9dQNLWfe95vtd0wMOB8Hofh5H0ex3DLSLVAFxt+jTcV8g+k8JPFxrDPmfs-P9sAKxJ5ALs5-pvfBTjR8VkAA)

Reply:
Cool! TS is a blast 😄